test/functional/wallet_coinselection.py

```diff
@@ -0,0 +1,261 @@
+            w1.lockunspent(False, [tx for tx in w1.listunspent() if tx['txid'] == txid])
+            return fee_total, max_fee
+
+        # in all the scenarios below the best solution would be to take one big coin and create a change output
+        best_fee = (42 + 68 + 31) * fee_rate / COIN
+
+        # 0) no exact match is possible, the big coin is closest to target + MIN_CHANGE,
+        # knapsack always finds the best solution
+        fee_total, max_fee = select_coins(big_coin=1.01051, target=1.00000001, iterations=100)
+        assert_approx(max_fee, best_fee)
+        assert_approx(fee_total, best_fee * 100)
+
+        # 1) knapsack latches to the exact match solution and can't find the best solution
+        # SRD can find a solution better than knapsack if draws the big coin in the first 21 tries
+        # chances for SRD to find it are roughly 2*(21/102) = ±40%.
+        # Chances are doubled due to coinselection run twice with and without APS enabled
+        fee_total, max_fee = select_coins(big_coin=1.01051, target=1.0, iterations=100)
+        knapsack_fee = (42 + 31 + 148*20) * fee_rate / COIN
+        assert_approx(max_fee, knapsack_fee)
+        # total fee is lower due to lucky SRD solutions
+        assert_greater_than(knapsack_fee * 100, fee_total)
+        assert_greater_than(fee_total, best_fee * 100)
+
+        # 2) no exact match, effective value of the big coin is smaller than target + MIN_CHANGE
+        # BnB can find the same solution from previous test but dropping the change output to fees
+        # SRD still occasionally find better solution
+        fee_total, max_fee = select_coins(big_coin=1.01031, target=1.00000001, iterations=100)
+        dropped_to_fees = 31 * fee_rate
+        bnb_fee = ((42 + 148*20) * fee_rate + dropped_to_fees) / COIN
+        assert_approx(max_fee, bnb_fee)
+        assert_greater_than(bnb_fee * 100, fee_total)
+        assert_greater_than(fee_total, best_fee * 100)
+
+        # 3) target is too far away for BnB to find a solution
+        # effective value of the big coin is smaller than target + MIN_CHANGE
+        #  * BnB can't find a solution as the distance from target is farther than cost of change
+        #    6061550 + 5000000*19-(42+148*20)*350-99999319 = 11531
+        #    cost_of_change = 11530 = 31*350 + 68*10
+        #  * knapsack finds closest solution to target + MIN_CHANGE with 21 small coins
+        #  * SRD still occasionally find better solution
+        fee_total, max_fee = select_coins(big_coin=1.01031, target=0.99999319, iterations=100)
+        srd_fee = (41 + 147*21 + 31) * fee_rate / COIN
+        assert_approx(max_fee, srd_fee)
+        assert_greater_than(srd_fee * 100, fee_total)
+        assert_greater_than(fee_total, best_fee * 100)
+
+        # 4) knapsack latches to exact match + MIN_CHANGE
+        # SRD still occasionally find better solution
+        fee_total, max_fee = select_coins(big_coin=1.00051, target=0.99, iterations=100)
```

Comment:
It's a bit hard for me to follow how the exact values were picked, since both the `big_coin` and `target` are changing. I haven't tried to step by step reproduce each calculation in this first review, and I do see that many of your test cases do explain some of the calculation, but I just noticed that this one for example does not.

Since it's not trivial for me to follow, I kinda expect that other reviewers might also need a bit help. :grin:

Reply:
This is basically the same as test case "1)" in this section, but all the values decreased by MIN_CHANGE. The idea is to  demonstrate that Knapsack latches not only to down-to-satoshi matches. I'll improve the comment